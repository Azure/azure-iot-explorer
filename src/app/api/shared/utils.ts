/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { createHmac } from 'crypto';
import { IoTHubConnectionSettings } from '../services/devicesService';
import { LIST_PLUG_AND_PLAY_DEVICES, SAS_EXPIRES_MINUTES } from '../../constants/devices';
import DeviceQuery, { QueryClause, ParameterType, OperationType } from '../models/deviceQuery';
import { RepoConnectionSettings } from '../services/digitalTwinsModelService';
import { AppEnvironment } from '../../constants/shared';

const MILLISECONDS_PER_SECOND = 1000;
const SECONDS_PER_MINUTE = 60;

export const enum PnPQueryPrefix {
    HAS_CAPABILITY_MODEL = 'HAS_CAPABILITYMODEL',
    HAS_INTERFACE = 'HAS_INTERFACE'
}

export interface GenerateSasTokenParameters {
    /**
     * URI prefix (by segment) of the endpoints that can be accessed with this token, starting with host name of the IoT hub (no protocol). For example, myHub.azure-devices.net/devices/device1
     */
    resourceUri: string;

    /**
     * The key to use for the token.
     */
    key: string;

    /**
     * Expiration time in minutes. Defaults to 5.
     */
    expiration?: number;

    /**
     * The name of the key being used. Omit if the token refers to device-registry credentials.
     */
    keyName?: string;
}

export const generateSasToken = (parameters: GenerateSasTokenParameters) => {
    let token = null;
    if (!!parameters.resourceUri && !!parameters.key) {
        const encodedUri = encodeURIComponent(parameters.resourceUri);

        const expires = Math.ceil((Date.now() / MILLISECONDS_PER_SECOND) + (parameters.expiration || SAS_EXPIRES_MINUTES) * SECONDS_PER_MINUTE);
        const toSign = encodedUri + '\n' + expires;

        const hmac = createHmac('sha256', new Buffer(parameters.key, 'base64'));
        hmac.update(toSign);
        const base64UriEncoded = encodeURIComponent(hmac.digest('base64'));

        token = `SharedAccessSignature sr=${encodedUri}&sig=${base64UriEncoded}&se=${expires}${parameters.keyName ? '&skn=' + parameters.keyName : ''}`;
    }
    return token;
};

export const generatePnpSasToken = (repositoryId: string, audience: string, secret: string, keyName: string) => {
    const now = new Date();
    const ms = 1000;
    const expiry = (now.setDate(now.getDate() + 1) / ms).toFixed(0);
    const encodedServiceEndpoint = encodeURIComponent(audience);
    const encodedRepoId = encodeURIComponent(repositoryId);
    const signature = [encodedRepoId, encodedServiceEndpoint, expiry].join('\n').toLowerCase();
    const sigUTF8 = new Buffer(signature, 'utf8');
    const secret64bit = new Buffer(secret, 'base64');
    const hmac = createHmac('sha256', secret64bit);
    hmac.update(sigUTF8);
    const hash = encodeURIComponent(hmac.digest('base64'));
    return `SharedAccessSignature sr=${encodedServiceEndpoint}&sig=${hash}&se=${expiry}&skn=${keyName}&rid=${repositoryId}`;
};

export const getRepoConnectionInfoFromConnectionString = (connectionString: string): RepoConnectionSettings => {
    const connectionObject: RepoConnectionSettings = {};
    connectionString.split(';')
    .forEach((segment: string) => {
        const keyValue = segment.split('=');
        switch (keyValue[0]) {
            case 'HostName':
            connectionObject.hostName = keyValue[1];
            break;
            case 'SharedAccessKeyName':
            connectionObject.sharedAccessKeyName = keyValue[1];
            break;
            case 'SharedAccessKey':
            connectionObject.sharedAccessKey = keyValue[1];
            break;
            case 'RepositoryId':
            connectionObject.repositoryId = keyValue[1];
            break;
            default:
            // we don't use other parts of connection string
            break;
        }
    });

    return connectionObject;
};

export const getConnectionInfoFromConnectionString = (connectionString: string): IoTHubConnectionSettings => {
    const connectionObject: IoTHubConnectionSettings = {};
    connectionString.split(';')
    .forEach((segment: string) => {
        const keyValue = segment.split('=');
        switch (keyValue[0]) {
            case 'HostName':
            connectionObject.hostName = keyValue[1];
            break;
            case 'SharedAccessKeyName':
            connectionObject.sharedAccessKeyName = keyValue[1];
            break;
            case 'SharedAccessKey':
            connectionObject.sharedAccessKey = keyValue[1];
            break;
            default:
            // we don't use other parts of connection string
            break;
        }
    });

    return connectionObject;
};

export const buildQueryString = (query: DeviceQuery) => {
    return query ?
        `${LIST_PLUG_AND_PLAY_DEVICES} ${queryToString(query)}` :
        LIST_PLUG_AND_PLAY_DEVICES;
};

export const queryToString = (query: DeviceQuery) => {
    if (query.deviceId) {
        return `WHERE STARTSWITH(devices.deviceId, '${query.deviceId}')`;
    }
    const clauseQuery = clauseListToString(query.clauses && query.clauses.filter(clause => !!clause.parameterType));
    if ('' !== clauseQuery) {
        return `WHERE (${clauseQuery })`;
    }
    return '';
};

export const clauseListToString = (clauses: QueryClause[]) => {
    return clauses && clauses.map(clause => clauseToString(clause)).join(' AND ') || '';
};

export const clauseToString = (clause: QueryClause) => {
    switch (clause.parameterType) {
        case ParameterType.capabilityModelId:
            return toPnPClause(PnPQueryPrefix.HAS_CAPABILITY_MODEL, clause.value);
        case ParameterType.interfaceId:
            return toPnPClause(PnPQueryPrefix.HAS_INTERFACE, clause.value);
        case ParameterType.edge:
            return toEdgeClause(clause.parameterType, clause.value);
        case ParameterType.status:
            return toDeviceStatusClause(clause.parameterType, clause.value);
        default:
            return clauseItemToString(clause.parameterType, clause.operation, clause.value);
    }
};

export const clauseItemToString = (fieldName: string, operation: OperationType, value: unknown) => {
    if (!fieldName || !operation || !value) {
        return;
    }
    return `${fieldName} ${operation} ${escapeValue(value as string)}`;
};

export const escapeValue = (value: string) => {
    const isString = new RegExp(/[^0-9.]/);
    const hasQuotes = new RegExp(/^'.*'$/);
    if (value.match(isString) && !value.match(hasQuotes)) {
        return `'${value}'`;
    }
    return value;
};

export const toPnPClause = (pnpFunctionName: string, value: string): string => {
    const urnVersionRegEx = new RegExp(/:[0-9]+$/);
    if (urnVersionRegEx.test(value)) {
        const urnVersion = urnVersionRegEx.exec(value)[0].replace(':', '');
        const urnName = value.replace(urnVersionRegEx, '');

        return `${pnpFunctionName}(${escapeValue(urnName)}, ${urnVersion})`;
    }
    return `${pnpFunctionName}('${value}')`; // when provided value is not urn with version, pass it in as string
};

export const toEdgeClause = (edgeFunctionName: string, value: string): string => {
    return `${edgeFunctionName}=${value === 'true' ? true : false}`;
};

export const toDeviceStatusClause = (edgeFunctionName: string, value: string): string => {
    return `${edgeFunctionName}='${value}'`;
};

export const isAppInElectron = () => {
    return process.env.NODE_ENV === AppEnvironment.DevelopmentElectron || process.env.NODE_ENV === AppEnvironment.ProdElectron;
};

export const isAppInDevelopmentMode = () => {
    return process.env.NODE_ENV === AppEnvironment.DevelopmentElectron || process.env.NODE_ENV === AppEnvironment.DevelopmentHosted;
};
