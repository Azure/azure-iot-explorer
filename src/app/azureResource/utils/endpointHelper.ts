/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/

// import { AzureResourceEndpointSet } from '../models/azureResourceEndpointSet';

// export const GetAzureResourceEndpointSetFromHostName = (hostName: string): AzureResourceEndpointSet => {
//     throw new Error('not implemented');
// };
