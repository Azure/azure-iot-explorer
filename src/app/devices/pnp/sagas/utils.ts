import { ModelIdCasingNotMatchingException } from '../../../shared/utils/exceptions/modelIdCasingNotMatchingException';
import { ModelDefinition } from '../../../api/models/modelDefinition';
import { ModelRepositoryConfiguration } from '../../../shared/modelRepository/state';
import { REPOSITORY_LOCATION_TYPE } from '../../../constants/repositoryLocationTypes';

export const getSplitInterfaceId = (fullName: string) => {
    // when component definition is inline, interfaceId is compose of parent file name and inline schema id concatenated with a slash
    return fullName.split('/');
};

export const getFlattenedModel = (model: ModelDefinition, splitInterfaceId: string[]) => {
    if (splitInterfaceId.length === 1) {
        return model;
    }
    else {
        // for inline component, the flattened model is defined under contents array's with matching schema @id
        const components = model.contents.filter((content: any) => // tslint:disable-line: no-any
            content['@type'] === 'Component' && typeof content.schema !== 'string' && content.schema['@id'] === splitInterfaceId[1]);
        return components[0];
    }
};

export const checkModelIdCasing = (model: ModelDefinition, id: string) => {
    if (model['@id'] !== id) {
        throw new ModelIdCasingNotMatchingException();
    }
};

export const getLocationSettingValue = (locations: ModelRepositoryConfiguration[], type: REPOSITORY_LOCATION_TYPE): string => {
    const filteredValue = locations.filter(location => location.repositoryLocationType === type)?.[0]?.value || '';
    return filteredValue.replace(/\/$/, ''); // remove trailing slash
};

export const getDmrParams = (path: string, interfaceId: string): {folderPath: string, fileName: string} => {
    // convert dtmi name to follow drm convention
    // for example: dtmi:com:example:Thermostat;1 -> dtmi/com/example/thermostat-1.json
    const fullPath = `${path.replace(/\/$/, '')}/${interfaceId.toLowerCase().replace(/:/g, '/').replace(';', '-')}.json`;
    // folderPath will be converted to for example: ${path}/dtmi/com/example, and fileName will be thermostat-1.json
    return {folderPath: fullPath.substring(0, fullPath.lastIndexOf('/')), fileName: fullPath.substring(fullPath.lastIndexOf('/') + 1, fullPath.length)};
};
