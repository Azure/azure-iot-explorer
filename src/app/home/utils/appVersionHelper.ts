/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as semver from 'semver';
export const isNewReleaseVersionHigher = (releaseVersion: string, appVersion: string) => {
    return semver.gt(releaseVersion, appVersion);
};
