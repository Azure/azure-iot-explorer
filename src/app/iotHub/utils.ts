/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/

export const getShortHostName = (name: string) => {
    return name && name.replace(/\..*/, '');
};
