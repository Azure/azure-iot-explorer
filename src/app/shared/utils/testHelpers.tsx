/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { LocalizationContextProvider } from '../contexts/localizationContext';
import configureStore from '../../../app/shared/redux/store/configureStore';

export const testWithLocalizationContext = (Target: JSX.Element, enzymeWrapper: any = shallow) => { // tslint:disable-line:no-any
    const outerWrapper = shallow(Target);
    const Children = outerWrapper.props().children;

    return enzymeWrapper(<Children t={jest.fn((value: string) => value)}/>);
};

export const testWithLocalizationContextAndErrorBoundary = (Target: JSX.Element, enzymeWrapper: any = shallow) => { // tslint:disable-line:no-any
    const outerWrapper = shallow(Target);
    const Children = outerWrapper.props().children.props.children;

    return enzymeWrapper(<Children t={jest.fn()}/>);
};

export const testSnapshot = (Target: JSX.Element) => {
    const outerWrapper = shallow(Target);
    const Children = outerWrapper.props().children;
    expect(shallow(<Children t={jest.fn((value: string) => value)}/>)).toMatchSnapshot();
};

export const mountWithLocalization = (Target: JSX.Element, connectToStore: boolean = false, useMemoryRouter: boolean = false, routerInitialEntries = ['/']) => {
    let wrapper = (
        <LocalizationContextProvider value={{t: jest.fn((value: string) => value)}}>
            {Target}
        </LocalizationContextProvider>
    );

    if (connectToStore) {
        wrapper = (
            <Provider store={configureStore()}>
                {wrapper}
            </Provider>
        );
    }

    if ( useMemoryRouter ) {
        wrapper = (
            <MemoryRouter initialEntries={routerInitialEntries} keyLength={0}>
                {wrapper}
            </MemoryRouter>
        );
    }

    return mount(wrapper);
};
