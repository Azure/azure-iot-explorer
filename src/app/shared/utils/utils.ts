/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
export const generateKey = (generator: (byteArray: Uint8Array) => void = randomValueGenerator): string => {
    const byteLength = 32;
    const byteArray = new Uint8Array(byteLength);
    generator(byteArray);

    let bytes = '';
    for (let i = 0; i < byteLength; i++) {
        bytes += String.fromCharCode(byteArray[i]);
    }

    return btoa(bytes);
};

export const randomValueGenerator = (byteArray: Uint8Array) => {
    // tslint:disable-next-line:no-any
    const windowAny = window as any;  // msCrypto is not typed.
    const crypto = windowAny.crypto || windowAny.msCrypto;
    crypto.getRandomValues(byteArray);
};

export const validateKey = (key: string): boolean => {
    const pattern = new RegExp('^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{4}|[A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)$');
    return pattern.test(key);
};

export const validateThumbprint = (key: string): boolean => {
    const pattern = new RegExp('^[A-Fa-f0-9]{40}$');
    return pattern.test(key);
};

export const validateDeviceId = (key: string): boolean => {
    const pattern = new RegExp('^[A-Za-z0-9-:.+%_#*?!(),=@;$\']{0,128}$');
    return pattern.test(key);
};

export const validateModuleIdentityName = (key: string): boolean => {
    const pattern = new RegExp('^[A-Za-z0-9-:.+%_#*?!(),=@;$\']{0,128}$');
    return pattern.test(key);
};
