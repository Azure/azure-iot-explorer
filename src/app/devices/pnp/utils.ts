/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import { getDeviceIdFromQueryString, getModuleIdentityIdFromQueryString } from '../../shared/utils/queryStringHelper';
import { ROUTE_PARAMS } from '../../constants/routes';
import { getDeviceTwinAction, getModuleTwinAction } from './actions';
import { ModelDefinition } from '../../api/models/modelDefinition';
import { ComponentToModelId, JsonSchemaAdaptor } from '../../shared/utils/jsonSchemaAdaptor';
import { DEFAULT_COMPONENT_FOR_DIGITAL_TWIN } from '../../constants/devices';

export const getBackUrl = (path: string, search: string) => {
    const deviceId = getDeviceIdFromQueryString(search);
    const moduleId = getModuleIdentityIdFromQueryString(search);
    let url = `${path}/?${ROUTE_PARAMS.DEVICE_ID}=${encodeURIComponent(deviceId)}`;
    if (moduleId) {
        url += `&${ROUTE_PARAMS.MODULE_ID}=${encodeURIComponent(moduleId)}`;
    }
    return url;
};

export const dispatchGetTwinAction = (search: string, dispatch: (action: any) => void) => {  // tslint:disable-line:no-any
    const deviceId = getDeviceIdFromQueryString(search);
    const moduleId = getModuleIdentityIdFromQueryString(search);
    if (moduleId) {
        dispatch(getModuleTwinAction.started({deviceId, moduleId}));
    }
    else {
        dispatch(getDeviceTwinAction.started(deviceId));
    }
};

export const getComponentNameAndInterfaceIdArray = (modelDefinition: ModelDefinition): ComponentToModelId[] => {
    if (!modelDefinition) {
        return [];
    }
    const jsonSchemaAdaptor = new JsonSchemaAdaptor(modelDefinition);
    const components = jsonSchemaAdaptor.getComponentNameToModelIdMapping();
    // check if model contains no-component items
    if (jsonSchemaAdaptor.getNonWritableProperties().length +
        jsonSchemaAdaptor.getWritableProperties().length +
        jsonSchemaAdaptor.getCommands().length +
        jsonSchemaAdaptor.getTelemetry().length > 0) {
        components.unshift({
            componentName: DEFAULT_COMPONENT_FOR_DIGITAL_TWIN,
            modelId: modelDefinition['@id']
        });
    }
    return components;
};
