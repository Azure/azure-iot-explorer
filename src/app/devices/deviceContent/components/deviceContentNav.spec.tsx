/***********************************************************
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License
 **********************************************************/
import 'jest';
import * as React from 'react';
import { Nav } from 'office-ui-fabric-react/lib/Nav';
import DeviceContentNavComponent, { DeviceContentNavDataProps, DeviceContentNavDispatchProps, NAV_LINK_ITEMS_NONPNP, NAV_LINK_ITEMS_PNP, NAV_LINK_ITEMS_NONPNP_NONEDGE } from './deviceContentNav';
import { mountWithLocalization, testSnapshot } from '../../../shared/utils/testHelpers';

describe('components/devices/deviceContentNav', () => {

    const setInterfaceId = jest.fn();
    const getComponent = (overrides = {}) => {
        const routeProps = {
            history: jest.fn() as any, // tslint:disable-line:no-any
            location: jest.fn() as any, // tslint:disable-line:no-any
            match: jest.fn() as any, // tslint:disable-line:no-any
        };

        const navDataProps: DeviceContentNavDataProps = {
            deviceId: 'test',
            interfaceIds: [],
            isEdgeDevice: true,
            isLoading: false,
            isPnPDevice: false,
            selectedInterface: '',
        };

        const navDispatchProps: DeviceContentNavDispatchProps = {
            setInterfaceId
        };

        const props = {
            ...navDispatchProps,
            ...navDataProps,
            ...routeProps
            ...overrides,
        };

        return <DeviceContentNavComponent {...props} />;
    };

    it('matches snapshot when there device is not pnp', () => {
        testSnapshot(getComponent());
        const wrapper = mountWithLocalization(getComponent());
        const navigation = wrapper.find(Nav);
        expect(navigation.props().groups[0].links.length).toEqual(NAV_LINK_ITEMS_NONPNP.length);
    });

    it('shows non-pnp nav when component is loading', () => {
        const wrapper = mountWithLocalization(getComponent({isLoading: true}));

        const navigation = wrapper.find(Nav);
        expect(navigation.props().groups[0].links.length).toEqual(NAV_LINK_ITEMS_NONPNP.length);
    });

    it('shows non-pnp non-edge nav if device is not edge', () => {
        const wrapper = mountWithLocalization(getComponent({isEdgeDevice: false}));

        const navigation = wrapper.find(Nav);
        expect(navigation.props().groups[0].links.length).toEqual(NAV_LINK_ITEMS_NONPNP_NONEDGE.length);
    });

    it('show non-pnp nav and pnp nav when device is pnp', () => {
        const interfaceId = 'urn:azureiot:com:DeviceInformation:1';
        const interfaceIds = [interfaceId];
        const wrapper = mountWithLocalization(getComponent({isPnPDevice: true, interfaceIds}));

        const navigation = wrapper.find(Nav);
        expect(navigation.props().groups[0].links.length).toEqual(NAV_LINK_ITEMS_NONPNP.length);
        expect(navigation.props().groups[1].links.length).toEqual(interfaceIds.length);
        expect(navigation.props().groups[1].links[0].links.length).toEqual(NAV_LINK_ITEMS_PNP.length);

        navigation.props().groups[1].links[0].links[0].onClick(undefined, {name: '', url: '', parentId: interfaceId});
        expect(setInterfaceId).toBeCalledWith(interfaceId);
    });
});
