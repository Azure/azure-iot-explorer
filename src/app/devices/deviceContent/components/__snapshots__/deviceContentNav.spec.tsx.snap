// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/devices/deviceContentNav matches snapshot when there device is not pnp 1`] = `
<div
  className="view-scroll"
>
  <div
    role="navigation"
  >
    <StyledNavBase
      groups={
        Array [
          Object {
            "links": Array [
              Object {
                "key": "identity",
                "name": "deviceContent.navBar.identity",
                "url": "#undefined/identity/?id=test",
              },
              Object {
                "key": "twin",
                "name": "deviceContent.navBar.twin",
                "url": "#undefined/twin/?id=test",
              },
              Object {
                "key": "events",
                "name": "deviceContent.navBar.events",
                "url": "#undefined/events/?id=test",
              },
              Object {
                "key": "methods",
                "name": "deviceContent.navBar.methods",
                "url": "#undefined/methods/?id=test",
              },
              Object {
                "key": "cloudToDeviceMessage",
                "name": "deviceContent.navBar.cloudToDeviceMessage",
                "url": "#undefined/cloudToDeviceMessage/?id=test",
              },
            ],
            "name": "deviceContent.navBar.nonpnp",
          },
        ]
      }
      onLinkExpandClick={[Function]}
      onRenderGroupHeader={[Function]}
    />
  </div>
</div>
`;
